In [26]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns

In [27]:
import sys

sys.path.append("../")

from src.eda.loader import DataLoader

loader = DataLoader("../data/raw/complaints.csv")

df = loader.get_data()
df.head()

c:\Users\hp\Desktop\Projects\ML\KAIM\week-6\fincomplain-ragbot\notebooks\..\src\eda\loader.py:13: DtypeWarning: Columns (16) have mixed types. Specify dtype option on import or set low_memory=False.
  self.df = pd.read_csv(self.filepath)


Date received                                            Product  \
0    2025-06-20  Credit reporting or other personal consumer re...   
1    2025-06-20                                    Debt collection   
2    2025-06-20  Credit reporting or other personal consumer re...   
3    2025-06-20  Credit reporting or other personal consumer re...   
4    2025-06-20  Credit reporting or other personal consumer re...   

               Sub-product                                 Issue  \
0         Credit reporting  Incorrect information on your report   
1  Telecommunications debt     Attempts to collect debt not owed   
2         Credit reporting           Improper use of your report   
3         Credit reporting           Improper use of your report   
4         Credit reporting  Incorrect information on your report   

                                       Sub-issue Consumer complaint narrative  \
0            Information belongs to someone else                          NaN   
1                              Debt is not yours                          NaN   
2  Reporting company used your report improperly                          NaN   
3  Reporting company used your report improperly                          NaN   
4                       Account status incorrect                          NaN   

                             Company public response  \
0                                                NaN   
1  Company can't verify or dispute the facts in t...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                        Company State ZIP code Tags  \
0           Experian Information Solutions Inc.    FL    32092  NaN   
1  Eastern Account Systems of Connecticut, Inc.    FL    342XX  NaN   
2        TRANSUNION INTERMEDIATE HOLDINGS, INC.    AZ    85225  NaN   
3           Experian Information Solutions Inc.    AZ    85225  NaN   
4           Experian Information Solutions Inc.    IL    60628  NaN   

  Consumer consent provided? Submitted via Date sent to company  \
0                        NaN           Web           2025-06-20   
1                        NaN           Web           2025-06-20   
2                        NaN           Web           2025-06-20   
3                        NaN           Web           2025-06-20   
4                        NaN           Web           2025-06-20   

  Company response to consumer Timely response? Consumer disputed?  \
0                  In progress              Yes                NaN   
1      Closed with explanation              Yes                NaN   
2                  In progress              Yes                NaN   
3                  In progress              Yes                NaN   
4                  In progress              Yes                NaN   

   Complaint ID  
0      14195687  
1      14195688  
2      14195689  
3      14195690  
4      14195692

### EDA

In [29]:
df.info()
df_saved = df.copy()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9609797 entries, 0 to 9609796
Data columns (total 18 columns):
 #   Column                        Dtype 
---  ------                        ----- 
 0   Date received                 object
 1   Product                       object
 2   Sub-product                   object
 3   Issue                         object
 4   Sub-issue                     object
 5   Consumer complaint narrative  object
 6   Company public response       object
 7   Company                       object
 8   State                         object
 9   ZIP code                      object
 10  Tags                          object
 11  Consumer consent provided?    object
 12  Submitted via                 object
 13  Date sent to company          object
 14  Company response to consumer  object
 15  Timely response?              object
 16  Consumer disputed?            object
 17  Complaint ID                  int64 
dtypes: int64(1), object(17)
memory usage: 1.3+

In [30]:
df.isnull().sum().sort_values(ascending=False)

Tags                            8981029
Consumer disputed?              8841498
Consumer complaint narrative    6629041
Company public response         4770207
Consumer consent provided?      1649561
Sub-issue                        839522
Sub-product                      235295
State                             54516
ZIP code                          30228
Company response to consumer         20
Issue                                 6
Date received                         0
Product                               0
Company                               0
Date sent to company                  0
Submitted via                         0
Timely response?                      0
Complaint ID                          0
dtype: int64

In [31]:
df['Product'].value_counts()

Product
Credit reporting or other personal consumer reports                             4834855
Credit reporting, credit repair services, or other personal consumer reports    2163857
Debt collection                                                                  799197
Mortgage                                                                         422254
Checking or savings account                                                      291178
Credit card                                                                      226686
Credit card or prepaid card                                                      206369
Money transfer, virtual currency, or money service                               145066
Credit reporting                                                                 140429
Student loan                                                                     109717
Bank account or service                                                           86205
Vehicle loan or lease   

### Preprocessing

#### Product filtering

In [32]:
def map_product(product: str) -> str:
    product = product.lower()

    # CREDIT CARD
    if product in [
        "credit card",
        "credit card or prepaid card",
        "prepaid card"
    ]:
        return "Credit Card"

    # PERSONAL LOAN
    elif product in [
        "consumer loan",
        "payday loan, title loan, or personal loan",
        "payday loan, title loan, personal loan, or advance loan",
        "payday loan"
    ]:
        return "Personal Loan"

    # BNPL
    elif "buy now" in product:
        return "BNPL"

    # SAVINGS ACCOUNT
    elif product in [
        "checking or savings account",
        "bank account or service"
    ]:
        return "Savings Account"

    # MONEY TRANSFERS
    elif product in [
        "money transfers",
        "money transfer, virtual currency, or money service"
    ]:
        return "Money Transfers"

    # All others: discard
    else:
        return None


In [36]:
df = df_saved.copy()

In [37]:
df["Product"] = df["Product"].apply(map_product)
df = df[df["Product"].notna()]

In [38]:
df["Product"].value_counts()


Product
Credit Card        448335
Savings Account    377383
Money Transfers    150420
Personal Loan       84270
Name: count, dtype: int64

#### Consumer compliant narrative

In [39]:
count = df['Consumer complaint narrative'].dropna().str.split().str.len()

In [40]:
print(count.min())
print(count.max())
print(count.std())
print(count.median())
print(count.mean())

1
6469
224.97469644484403
137.0
205.14300633144674


In [47]:
import re

def clean_text(text: str) -> str:
    text = text.lower()
    text = re.sub(r"i am writing to (file|submit) a complaint", "", text)
    text = re.sub(r"[^a-zA-Z0-9\s\.\?]", " ", text)
    text = re.sub(r"\s+", " ", text)
    return text.strip()

In [42]:
df.head(10)

Date received          Product  \
334     2025-06-20      Credit Card   
351     2025-06-07  Savings Account   
437     2025-06-07  Savings Account   
595     2025-06-20      Credit Card   
661     2025-06-20      Credit Card   
687     2025-06-20      Credit Card   
795     2025-06-20      Credit Card   
808     2025-06-20  Money Transfers   
1025    2025-06-20  Savings Account   
1031    2025-06-20      Credit Card   

                                     Sub-product  \
334   General-purpose credit card or charge card   
351                             Checking account   
437                             Checking account   
595   General-purpose credit card or charge card   
661   General-purpose credit card or charge card   
687   General-purpose credit card or charge card   
795   General-purpose credit card or charge card   
808                             Virtual currency   
1025                            Checking account   
1031  General-purpose credit card or charge card   

                                                  Issue  \
334   Problem with a company's investigation into an...   
351                                 Managing an account   
437                                  Closing an account   
595   Problem with a company's investigation into an...   
661   Problem with a company's investigation into an...   
687   Problem with a company's investigation into an...   
795                Incorrect information on your report   
808                               Other service problem   
1025                                 Closing an account   
1031                              Getting a credit card   

                                              Sub-issue  \
334   Was not notified of investigation status or re...   
351                   Problem using a debit or ATM card   
437                         Company closed your account   
595   Was not notified of investigation status or re...   
661   Was not notified of investigation status or re...   
687   Was not notified of investigation status or re...   
795                            Account status incorrect   
808                                                 NaN   
1025             Funds not received from closed account   
1031        Card opened without my consent or knowledge   

     Consumer complaint narrative Company public response  \
334                           NaN                     NaN   
351                           NaN                     NaN   
437                           NaN                     NaN   
595                           NaN                     NaN   
661                           NaN                     NaN   
687                           NaN                     NaN   
795                           NaN                     NaN   
808                           NaN                     NaN   
1025                          NaN                     NaN   
1031                          NaN                     NaN   

                                     Company State ZIP code Tags  \
334                            EQUIFAX, INC.    TX    75104  NaN   
351                    WELLS FARGO & COMPANY    TX    75034  NaN   
437                    WELLS FARGO & COMPANY    NV    89115  NaN   
595      Experian Information Solutions Inc.    CA    92553  NaN   
661   TRANSUNION INTERMEDIATE HOLDINGS, INC.    TX    77346  NaN   
687                            EQUIFAX, INC.    TX    77346  NaN   
795   TRANSUNION INTERMEDIATE HOLDINGS, INC.    LA    71108  NaN   
808                              Block, Inc.    GA    30213  NaN   
1025                        TransferWise Ltd    DE    19901  NaN   
1031                           EQUIFAX, INC.    TX    75065  NaN   

     Consumer consent provided? Submitted via Date sent to company  \
334                         NaN           Web           2025-06-20   
351                         NaN      Referral           2025-06-09   
437                         NaN           Web           2025-06-07   
5

In [43]:

df = df[df['Consumer complaint narrative'].notna()]

In [44]:
df.dtypes

Date received                   object
Product                         object
Sub-product                     object
Issue                           object
Sub-issue                       object
Consumer complaint narrative    object
Company public response         object
Company                         object
State                           object
ZIP code                        object
Tags                            object
Consumer consent provided?      object
Submitted via                   object
Date sent to company            object
Company response to consumer    object
Timely response?                object
Consumer disputed?              object
Complaint ID                     int64
dtype: object

In [46]:
df['Consumer complaint narrative'].isnull().sum()

np.int64(0)

In [48]:
df = df[df['Consumer complaint narrative'].str.split().str.len() >= 1]
df['Consumer complaint narrative'] = df['Consumer complaint narrative'].apply(clean_text)
df.head()

Date received          Product  \
12237    2025-06-13      Credit Card   
12532    2025-06-13  Savings Account   
13280    2025-06-12      Credit Card   
13506    2025-06-12      Credit Card   
13955    2025-06-09      Credit Card   

                                      Sub-product  \
12237                           Store credit card   
12532                            Checking account   
13280  General-purpose credit card or charge card   
13506  General-purpose credit card or charge card   
13955  General-purpose credit card or charge card   

                                                 Issue  \
12237                            Getting a credit card   
12532                              Managing an account   
13280               Other features, terms, or problems   
13506             Incorrect information on your report   
13955  Problem with a purchase shown on your statement   

                                               Sub-issue  \
12237        Card opened without my consent or knowledge   
12532                           Deposits and withdrawals   
13280                                      Other problem   
13506                      Account information incorrect   
13955  Credit card company isn't resolving a dispute ...   

                            Consumer complaint narrative  \
12237  a xxxx xxxx card was opened under my name by a...   
12532  i made the mistake of using my wellsfargo debi...   
13280  dear cfpb i have a secured credit card with ci...   
13506  i have a citi rewards cards. the credit balanc...   
13955  b i am writing to dispute the following charge...   

                                 Company public response  \
12237  Company has responded to the consumer and the ...   
12532  Company has responded to the consumer and the ...   
13280  Company has responded to the consumer and the ...   
13506  Company has responded to the consumer and the ...   
13955  Company has responded to the consumer and the ...   

                     Company State ZIP code            Tags  \
12237         CITIBANK, N.A.    TX    78230   Servicemember   
12532  WELLS FARGO & COMPANY    ID    83815             NaN   
13280         CITIBANK, N.A.    NY    11220             NaN   
13506         CITIBANK, N.A.    IL    60067             NaN   
13955         CITIBANK, N.A.    TX    78413  Older American   

      Consumer consent provided? Submitted via Date sent to company  \
12237           Consent provided           Web           2025-06-13   
12532           Consent provided           Web           2025-06-13   
13280           Consent provided           Web           2025-06-13   
13506           Consent provided           Web           2025-06-12   
13955           Consent provided           Web           2025-06-09   

          Company response to consumer Timely response? Consumer disputed?  \
12237  Closed with non-monetary relief              Yes                NaN   
12532          Closed with explanation              Yes                NaN   
13280      Closed with monetary relief              Yes                NaN   
13506          Closed with explanation              Yes                NaN   
13955      Closed with monetary relief              Yes                NaN   

       Complaint ID  
12237      14069121  
12532      14061897  
13280      14047085  
13506      14040217  
13955      13968411

In [53]:
df["Product"].value_counts()

Product
Credit Card        197126
Savings Account    155204
Money Transfers     98685
Personal Loan       37341
Name: count, dtype: int64

In [52]:
df.shape

(488356, 18)

In [51]:
df.to_csv("../data/processed/filtered_compliants.csv", index=False)

**Exploratory Analysis Summary**

After conducting EDA on the CFPB complaint dataset and performing targeted preprocessing, the following key insights emerged:

*   The raw dataset contained over 8.9 million complaints, but more than **6.6 million** were missing usable narrative text.
    
*   We focused on **five relevant product categories**, but only four had sufficient volume:**Credit Card, Personal Loan, Savings Account, and Money Transfers.**
    
*   Using an if-else mapping function, we normalized product names to ensure consistency across categories.
    
    *   Credit Card 197,126 
    *   Savings Account 155,204 
    *   Money Transfers 98,685 
    *   Personal Loan 37,341
    
*   Text cleaning included removing boilerplate phrases, converting to lowercase, and normalizing whitespace and punctuation.
    
*   A wide variance in narrative length confirmed the need for chunking before semantic search.